pkg/alertmanager/operator.go

```diff
@@ -469,6 +637,336 @@ func (c *Operator) sync(ctx context.Context, key string) error {
+			return errors.Wrapf(
+				err,
+				"failed to check whether generated config secret already exists for Alertmanager %v in namespace %v",
+				am.Name,
+				am.Namespace,
+			)
+		}
+		_, err = sClient.Create(context.TODO(), generatedConfigSecret, metav1.CreateOptions{})
+		level.Debug(c.logger).Log("msg", "created generated config secret", "secretname", generatedConfigSecret.Name)
+
+	} else {
+		_, err = sClient.Update(context.TODO(), generatedConfigSecret, metav1.UpdateOptions{})
+		level.Debug(c.logger).Log("msg", "updated generated config secret", "secretname", generatedConfigSecret.Name)
+	}
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to create generated config secret for Alertmanager %v in namespace %v", am.Name, am.Namespace)
+	}
+
+	return nil
+}
+
+func (c *Operator) selectAlertManagerConfigs(am *monitoringv1.Alertmanager) (map[string]*monitoringv1.AlertmanagerConfig, error) {
+	namespaces := []string{}
+	// Selectors (<namespace>/<name>) might overlap. Deduplicate them along the keyFunc.
+	res := make(map[string]*monitoringv1.AlertmanagerConfig)
+
+	amConfigSelector, err := metav1.LabelSelectorAsSelector(am.Spec.AlertmanagerConfigSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// If 'AlertmanagerConfigNamespaceSelector' is nil, only check own namespace.
+	if am.Spec.AlertmanagerConfigNamespaceSelector == nil {
+		namespaces = append(namespaces, am.Namespace)
+	} else {
+		amConfigNSSelector, err := metav1.LabelSelectorAsSelector(am.Spec.AlertmanagerConfigNamespaceSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		namespaces, err = c.listMatchingNamespaces(amConfigNSSelector)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	level.Debug(c.logger).Log("msg", "filtering namespaces to select AlertmanagerConfigs from", "namespaces", strings.Join(namespaces, ","), "namespace", am.Namespace, "alertmanager", am.Name)
+
+	for _, ns := range namespaces {
+		cache.ListAllByNamespace(c.alrtCfgInf.GetIndexer(), ns, amConfigSelector, func(obj interface{}) {
+			k, ok := c.keyFunc(obj)
+			if ok {
+				res[k] = obj.(*monitoringv1.AlertmanagerConfig)
+			}
+		})
+	}
+
+	alertmanagerConfigs := []string{}
+	for k := range res {
+		alertmanagerConfigs = append(alertmanagerConfigs, k)
+	}
+	level.Debug(c.logger).Log("msg", "selected AlertmanagerConfigs", "aletmanagerconfigs", strings.Join(alertmanagerConfigs, ","), "namespace", am.Namespace, "prometheus", am.Name)
+
+	return res, nil
+}
+
+// listMatchingNamespaces lists all the namespaces that match the provided
+// selector.
+func (c *Operator) listMatchingNamespaces(selector labels.Selector) ([]string, error) {
+	var ns []string
+	err := cache.ListAll(c.nsAlrtInf.GetStore(), selector, func(obj interface{}) {
+		ns = append(ns, obj.(*v1.Namespace).Name)
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list namespaces")
+	}
+	return ns, nil
+}
+
+func (c *Operator) createOrUpdateTLSAssetSecret(am *monitoringv1.Alertmanager) error {
```

Comment:
I copied these functions for handling the tls assets from the prometheus controller, and might have tweaked slightly to fit.

Reply:
Check #3445

Reply:
Done.